src/backend/vulkan/src/physical_device.rs

```diff
@@ -12,162 +12,673 @@ use hal::{
+                mem::transmute::<_, &mut vk::BaseOutStructure>(features).p_next = ptr::null_mut();
+            }
+        }
+
+        unsafe {
+            null_p_next(&mut features.vulkan_1_2);
+            null_p_next(&mut features.descriptor_indexing);
+            null_p_next(&mut features.mesh_shader);
+            null_p_next(&mut features.imageless_framebuffer);
+        }
+
+        (device_properties, features)
+    }
 }
 
-impl fmt::Debug for PhysicalDevice {
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        fmt.write_str("PhysicalDevice")
+pub struct PhysicalDevice {
+    instance: Arc<RawInstance>,
+    pub handle: vk::PhysicalDevice,
+    known_memory_flags: vk::MemoryPropertyFlags,
+    device_info: PhysicalDeviceInfo,
+    device_features: PhysicalDeviceFeatures,
+}
+
+pub(crate) fn load_adapter(
+    instance: &Arc<RawInstance>,
+    device: vk::PhysicalDevice,
+) -> adapter::Adapter<Backend> {
+    let (device_info, device_features) = PhysicalDeviceInfo::load(instance, device);
+
+    let info = adapter::AdapterInfo {
+        name: unsafe {
+            CStr::from_ptr(device_info.properties.device_name.as_ptr())
+                .to_str()
+                .unwrap_or("Unknown")
+                .to_owned()
+        },
+        vendor: device_info.properties.vendor_id as usize,
+        device: device_info.properties.device_id as usize,
+        device_type: match device_info.properties.device_type {
+            ash::vk::PhysicalDeviceType::OTHER => adapter::DeviceType::Other,
+            ash::vk::PhysicalDeviceType::INTEGRATED_GPU => adapter::DeviceType::IntegratedGpu,
+            ash::vk::PhysicalDeviceType::DISCRETE_GPU => adapter::DeviceType::DiscreteGpu,
+            ash::vk::PhysicalDeviceType::VIRTUAL_GPU => adapter::DeviceType::VirtualGpu,
+            ash::vk::PhysicalDeviceType::CPU => adapter::DeviceType::Cpu,
+            _ => adapter::DeviceType::Other,
+        },
+    };
+
+    let physical_device = PhysicalDevice {
+        instance: instance.clone(),
+        handle: device,
+        known_memory_flags: vk::MemoryPropertyFlags::DEVICE_LOCAL
+            | vk::MemoryPropertyFlags::HOST_VISIBLE
+            | vk::MemoryPropertyFlags::HOST_COHERENT
+            | vk::MemoryPropertyFlags::HOST_CACHED
+            | vk::MemoryPropertyFlags::LAZILY_ALLOCATED,
+        device_info,
+        device_features,
+    };
+
+    let queue_families = unsafe {
+        instance
+            .inner
+            .get_physical_device_queue_family_properties(device)
+            .into_iter()
+            .enumerate()
+            .map(|(i, properties)| QueueFamily {
+                properties,
+                device,
+                index: i as u32,
+            })
+            .collect()
+    };
+
+    adapter::Adapter {
+        info,
+        physical_device,
+        queue_families,
     }
 }
 
-pub struct DeviceCreationFeatures {
```

Comment:
Renamed to `PhysicalDeviceFeatures` and a `PhysicalDeviceVulkan12Features` field is added